File: backend/services/document-service/app/services/formOption_service.go
package services

import (
	"capstone/app/repositories"
	"capstone/app/schemas"
	"context"
)

type FormOptionsService interface {
	GetFormOptions(ctx context.Context) (*schemas.FormOptionsResponse, error)
	GetFormOptionsByGroup(ctx context.Context, userId string) (*schemas.FormOptionsResponse, error)
}

type formOptionsService struct {
	repo repositories.FormOptionsRepository
}

func NewFormOptionsService(repo repositories.FormOptionsRepository) FormOptionsService {
	return &formOptionsService{repo: repo}
}

// GetFormOptions returns all form options (used for breadcrumbs / navigation).
func (s *formOptionsService) GetFormOptions(ctx context.Context) (*schemas.FormOptionsResponse, error) {
	services, err := s.repo.GetAllNested(ctx)
	if err != nil {
		return nil, err
	}

	docTypes, err := s.repo.GetDocumentTypes(ctx)
	if err != nil {
		return nil, err
	}

	return &schemas.FormOptionsResponse{
		Services:      services,
		DocumentTypes: docTypes,
	}, nil
}

// GetFormOptionsByGroup returns form options filtered by the user's group (used for upload form).
func (s *formOptionsService) GetFormOptionsByGroup(ctx context.Context, userId string) (*schemas.FormOptionsResponse, error) {
	services, err := s.repo.GetAllNestedByGroup(ctx, userId)
	if err != nil {
		return nil, err
	}

	docTypes, err := s.repo.GetDocumentTypes(ctx)
	if err != nil {
		return nil, err
	}

	return &schemas.FormOptionsResponse{
		Services:      services,
		DocumentTypes: docTypes,
	}, nil
}

File: frontend/src/services/formOption_service.ts
import axioHandler from '@/cores/axios'
import { formOptionsResponseSchema } from '@/dtos/formOption_dto'
import type { FormOptionsResponse } from '@/dtos/formOption_dto'

export const formOptionsService = {
    // Returns ALL form options (used for breadcrumbs / navigation)
    getAll: async (): Promise<FormOptionsResponse> => {
        try {
            const { data } = await axioHandler.get('/form-option')

            const parsed = formOptionsResponseSchema.safeParse(data)
            if (!parsed.success) {
                console.error('[formOptionsService] Zod validation failed:', parsed.error.format())
                throw new Error('Gagal memvalidasi data form options dari server.')
            }

            return parsed.data
        } catch (error) {
            console.error('[formOptionsService] getAll error:', error)
            throw error instanceof Error ? error : new Error('Failed to fetch form options.')
        }
    },

    // Returns form options filtered by user's group (used for upload form)
    getByGroup: async (): Promise<FormOptionsResponse> => {
        try {
            const { data } = await axioHandler.get('/form-option', { params: { scope: 'group' } })

            const parsed = formOptionsResponseSchema.safeParse(data)
            if (!parsed.success) {
                console.error('[formOptionsService] Zod validation failed:', parsed.error.format())
                throw new Error('Gagal memvalidasi data form options dari server.')
            }

            return parsed.data
        } catch (error) {
            console.error('[formOptionsService] getByGroup error:', error)
            throw error instanceof Error ? error : new Error('Failed to fetch group form options.')
        }
    },
}